File: examples/agent-host/python/sandbox_profile.py
"""SandboxProfile 的配置契约；不声称自己创建了进程、网络或 Secret 隔离。"""

from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class SandboxProfile:
    workspace_roots: tuple[str, ...]
    process_allowlist: tuple[str, ...] = ()
    network_allowlist: tuple[str, ...] = ()
    secret_names: tuple[str, ...] = ()
    max_runtime_seconds: int = 30

    def validate(self) -> None:
        if not self.workspace_roots or any(not Path(root).is_absolute() for root in self.workspace_roots):
            raise ValueError("Sandbox workspace_roots 必须是非空绝对路径")
        if any("/" in item or "\\" in item for item in self.process_allowlist):
            raise ValueError("process_allowlist 只能使用可执行文件名")
        if any(item == "*" or "://" in item for item in self.network_allowlist):
            raise ValueError("network_allowlist 不允许通配符或 URL")
        if any(not item or "\n" in item for item in self.secret_names):
            raise ValueError("secret_names 不能为空且不能包含换行")
        if self.max_runtime_seconds <= 0:
            raise ValueError("max_runtime_seconds 必须为正数")

    def allows_process(self, executable: str) -> bool:
        return Path(executable).name in self.process_allowlist

    def allows_network(self, host: str) -> bool:
        return host in self.network_allowlist

    def may_log_secret(self, _name: str) -> bool:
        return False

File: examples/agent-host/typescript/sandbox-profile.ts
export type SandboxProfileConfig = {
  workspaceRoots: string[]; processAllowlist?: string[]; networkAllowlist?: string[]
  secretNames?: string[]; maxRuntimeSeconds?: number
}

export class SandboxProfile {
  readonly config: Required<SandboxProfileConfig>
  constructor(config: SandboxProfileConfig) {
    this.config = {
      workspaceRoots: config.workspaceRoots, processAllowlist: config.processAllowlist ?? [],
      networkAllowlist: config.networkAllowlist ?? [], secretNames: config.secretNames ?? [],
      maxRuntimeSeconds: config.maxRuntimeSeconds ?? 30,
    }
  }
  validate(): void {
    if (this.config.workspaceRoots.length === 0 || this.config.workspaceRoots.some((root) => !root.startsWith("/"))) throw new Error("Sandbox workspaceRoots 必须是非空绝对路径")
    if (this.config.processAllowlist.some((item) => /[\\/]/u.test(item))) throw new Error("processAllowlist 只能使用可执行文件名")
    if (this.config.networkAllowlist.some((item) => item === "*" || item.includes("://"))) throw new Error("networkAllowlist 不允许通配符或 URL")
    if (this.config.secretNames.some((item) => !item || /\n/u.test(item))) throw new Error("secretNames 非法")
    if (this.config.maxRuntimeSeconds <= 0) throw new Error("maxRuntimeSeconds 必须为正数")
  }
  allowsProcess(executable: string): boolean { return this.config.processAllowlist.includes(executable.split(/[\\/]/u).at(-1)!) }
  allowsNetwork(host: string): boolean { return this.config.networkAllowlist.includes(host) }
  mayLogSecret(_name: string): boolean { return false }
}
